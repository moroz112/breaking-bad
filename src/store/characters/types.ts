import { RequestState } from '../types';

export interface ICharacter {
  id: number;
  name: string;
  birthday: string;
  occupation: string[];
  img: string;
  status: string;
  nickname: string;
  appearance: string[];
  portrayed: string;
  category: string[];
}

export interface ICharactersState {
  currentCharacter: ICharacter | null;
  characterByIdGetState: RequestState | null;
  characterByIdGetError: string | null;
}

export const CHARACTER_GET_ONE = 'CHARACTER_GET_ONE';
export const CHARACTER_GET_ONE_SUCCESS = 'CHARACTER_GET_ONE_SUCCESS';
export const CHARACTER_GET_ONE_ERROR = 'CHARACTER_GET_ONE_ERROR';

// Get One Episode
export interface CharacterGetOneAction {
  type: typeof CHARACTER_GET_ONE;
  payload: string;
}

export interface CharacterGetOneSuccessAction {
  type: typeof CHARACTER_GET_ONE_SUCCESS;
  payload: ICharacter;
}

export interface CharacterGetOneErrorAction {
  type: typeof CHARACTER_GET_ONE_ERROR;
  payload: any;
}

export type CharactersActionTypes =
  | CharacterGetOneAction
  | CharacterGetOneSuccessAction
  | CharacterGetOneErrorAction;
