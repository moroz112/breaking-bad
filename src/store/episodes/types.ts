import { RequestState } from '../types';

export enum EpisodeSeries {
  BreakingBad = 'Breaking Bad',
  BetterCallSaul = 'Better Call Saul',
}

export interface IEpisode {
  episode_id: number;
  title: string;
  season: number;
  episode: number;
  air_date: string;
  characters: string[];
  series: EpisodeSeries;
}

export interface IEpisodesState {
  episodes: IEpisode[];
  episodesGetState: RequestState | null;
  episodesGetError: string | null;

  currentEpisode: IEpisode | null;
  episodeByIdGetState: RequestState | null;
  episodeByIdGetError: string | null;
}

export const EPISODES_GET = 'EPISODES_GET';
export const EPISODES_GET_SUCCESS = 'EPISODES_GET_SUCCESS';
export const EPISODES_GET_ERROR = 'EPISODES_GET_ERROR';

export const EPISODE_GET_ONE = 'EPISODE_GET_ONE';
export const EPISODE_GET_ONE_SUCCESS = 'EPISODE_GET_ONE_SUCCESS';
export const EPISODE_GET_ONE_ERROR = 'EPISODE_GET_ONE_ERROR';

export interface EpisodesGetAction {
  type: typeof EPISODES_GET;
  payload: EpisodeSeries;
}

export interface EpisodesGetSuccessAction {
  type: typeof EPISODES_GET_SUCCESS;
  payload: IEpisode[];
}

export interface EpisodesGetErrorAction {
  type: typeof EPISODES_GET_ERROR;
  payload: any;
}

export interface EpisodesGetOneAction {
  type: typeof EPISODE_GET_ONE;
  payload: number;
}

export interface EpisodesGetOneSuccessAction {
  type: typeof EPISODE_GET_ONE_SUCCESS;
  payload: IEpisode;
}

export interface EpisodesGetOneErrorAction {
  type: typeof EPISODE_GET_ONE_ERROR;
  payload: any;
}

export type EpisodesActionTypes =
  | EpisodesGetAction
  | EpisodesGetSuccessAction
  | EpisodesGetErrorAction
  | EpisodesGetOneAction
  | EpisodesGetOneSuccessAction
  | EpisodesGetOneErrorAction;
