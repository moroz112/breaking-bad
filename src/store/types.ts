import { ICharactersState } from './characters/types';
import { IEpisodesState } from './episodes/types';

export enum RequestState {
  Loading,
  Saving,
  Success,
  Error,
}

export interface AppStore {
  episodes: IEpisodesState;
  characters: ICharactersState;
}
