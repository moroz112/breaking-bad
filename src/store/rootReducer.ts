import { combineReducers } from 'redux';
import EpisodesReducer from './episodes/reducer';
import CharactersReducer from './characters/reducer';

const rootReducer = combineReducers({
  episodes: EpisodesReducer,
  characters: CharactersReducer,
});

export default rootReducer;
