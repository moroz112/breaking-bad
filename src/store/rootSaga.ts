import { all, fork } from 'redux-saga/effects';

import episodes from './episodes/sagas';
import characters from './characters/sagas';

// this is main root saga
export default function* root() {
  yield all([
      fork(episodes),
      fork(characters)]
  );
}
