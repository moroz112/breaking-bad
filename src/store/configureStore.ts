import thunk from 'redux-thunk';
import { createStore, applyMiddleware, compose } from 'redux';
import { createLogger } from 'redux-logger';
import createSagaMiddleware from 'redux-saga';

import rootReducer from './rootReducer';
import rootSaga from './rootSaga';

const sagaMiddleware = createSagaMiddleware();

const configureStore = (preloadedState: any) => {
  const enhancers = compose(
    applyMiddleware(thunk, createLogger(), sagaMiddleware),
  );
  const store = createStore(rootReducer, preloadedState, enhancers);
  return store;
};

const store = configureStore({});
sagaMiddleware.run(rootSaga);

export default store;
