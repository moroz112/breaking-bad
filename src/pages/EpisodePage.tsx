import clsx from 'clsx';
import Button from '../components/Button/Button';
import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useNavigate, useParams } from 'react-router-dom';
import { AppStore, RequestState } from '../store/types';
import { episodeGetOne } from '../store/episodes/actions';

export const humanizeString = (value: string) => {
  return value.split('_').join(' ')
}

const EpisodePage = () => {
  const { id } = useParams();
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { currentEpisode, episodeByIdGetState } = useSelector((store: AppStore) => store.episodes);

  useEffect(() => {
    if (id) {
      dispatch(episodeGetOne(Number(id)))
    }
  }, [dispatch, id]);

  if (episodeByIdGetState === RequestState.Loading || !currentEpisode) {
    return <>Loading in progress</>;
  }

  return (
    <div className="bg-white shadow overflow-hidden sm:rounded-lg max-w-2xl mx-auto mt-16">
      <div className="px-4 py-2 flex justify-between">
        <h3 className="text-lg leading-6 font-medium text-gray-900 mt-3">Episode info</h3>
        <Button onClick={() => navigate('/')}>Go To Season List</Button>
      </div>
      <div className="border-t border-gray-200">
        <dl>
          {Object.entries(currentEpisode)
            .filter(([key, val]) => key !== 'characters' && key !== 'episode_id')
            .map(([key, value], index) => (
              <div
                className='px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6'
                key={key}
              >
                <dt className="text-sm font-medium text-gray-500">{humanizeString(key)}</dt>
                <dd className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2">{value}</dd>
              </div>
            ))}
          <div className="bg-white px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6">
            <dt className="text-sm font-medium text-gray-500">Characters</dt>
            <dd className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2">
              <ul className="rounded-md">
                {currentEpisode.characters.map((character) => (
                  <li
                    className="pl-3 pr-4 py-3 flex items-center justify-between text-sm"
                    key={character}
                  >
                    <div className="w-0 flex-1 flex items-center">
                      <Link
                          to={`/character/${character}`}
                          className="font-medium text-indigo-600"
                      >
                        <span className="ml-2">{character}</span>
                      </Link>
                    </div>
                  </li>
                ))}
              </ul>
            </dd>
          </div>
        </dl>
      </div>
    </div>
  );
};

export default EpisodePage;
