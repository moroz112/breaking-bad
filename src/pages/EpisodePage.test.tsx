import React from 'react';
import renderer from 'react-test-renderer';
import { useDispatch, useSelector } from 'react-redux';
import { RequestState } from '../store/types';
import { useParams } from 'react-router-dom';
import EpisodePage, {
  humanizeString
} from './EpisodePage';

jest.mock('react-redux');
jest.mock('react-router');

const mockData = {
  episodeByIdGetState: RequestState.Success,
  currentEpisode: {
    air_date: '03-02-2015',
    characters: [
      'Jimmy McGill',
      'Mike Erhmantraut',
      'Kim Wexler',
      'Howard Hamlin',
      'Chuck McGill',
      'Nacho Varga',
    ],
    episode: '5',
    episode_id: 67,
    season: '1',
    series: 'Better Call Saul',
    title: 'Alpine Shepherd Boy',
  },
};

test('EpisodePage must render loading', () => {
  (useSelector as any).mockImplementationOnce(() => ({ episodeByIdGetState: RequestState.Loading }));
  (useParams as any).mockImplementationOnce(() => ({ id: '1' }));
  (useDispatch as any).mockImplementationOnce(() => () => {});
  const component = renderer.create(<EpisodePage />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('EpisodePage must render data', async () => {
  (useSelector as any).mockImplementationOnce(() => mockData);
  (useParams as any).mockImplementationOnce(() => ({ id: '1' }));
  (useDispatch as any).mockImplementationOnce(() => () => {});
  const component = renderer.create(<EpisodePage />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('humanize string works correctly', () => {
  const val = 'some_value';
  const result = humanizeString(val);
  const expected = 'some value';

  expect(expected).toEqual(result);
})
