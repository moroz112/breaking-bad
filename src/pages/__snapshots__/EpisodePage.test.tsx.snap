// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodePage must render data 1`] = `
<div
  className="bg-white shadow overflow-hidden sm:rounded-lg max-w-2xl mx-auto mt-16"
>
  <div
    className="px-4 py-2 flex justify-between"
  >
    <h3
      className="text-lg leading-6 font-medium text-gray-900 mt-3"
    >
      Episode info
    </h3>
    <button
      className="py-1 px-4 rounded my-2 mx-1 bg-gray-200 hover:bg-gray-400 text-black"
      onClick={[Function]}
    >
      Go To Season List
    </button>
  </div>
  <div
    className="border-t border-gray-200"
  >
    <dl>
      <div
        className="px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6"
      >
        <dt
          className="text-sm font-medium text-gray-500"
        >
          air date
        </dt>
        <dd
          className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
        >
          03-02-2015
        </dd>
      </div>
      <div
        className="px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6"
      >
        <dt
          className="text-sm font-medium text-gray-500"
        >
          episode
        </dt>
        <dd
          className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
        >
          5
        </dd>
      </div>
      <div
        className="px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6"
      >
        <dt
          className="text-sm font-medium text-gray-500"
        >
          season
        </dt>
        <dd
          className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
        >
          1
        </dd>
      </div>
      <div
        className="px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6"
      >
        <dt
          className="text-sm font-medium text-gray-500"
        >
          series
        </dt>
        <dd
          className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
        >
          Better Call Saul
        </dd>
      </div>
      <div
        className="px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6"
      >
        <dt
          className="text-sm font-medium text-gray-500"
        >
          title
        </dt>
        <dd
          className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
        >
          Alpine Shepherd Boy
        </dd>
      </div>
      <div
        className="bg-white px-4 py-5 sm:grid sm:grid-cols-3 sm:gap-4 sm:px-6"
      >
        <dt
          className="text-sm font-medium text-gray-500"
        >
          Characters
        </dt>
        <dd
          className="mt-1 text-sm text-gray-900 sm:mt-0 sm:col-span-2"
        >
          <ul
            className="rounded-md"
          >
            <li
              className="pl-3 pr-4 py-3 flex items-center justify-between text-sm"
            >
              <div
                className="w-0 flex-1 flex items-center"
              >
                <a
                  className="font-medium text-indigo-600"
                  onClick={[Function]}
                >
                  <span
                    className="ml-2"
                  >
                    Jimmy McGill
                  </span>
                </a>
              </div>
            </li>
            <li
              className="pl-3 pr-4 py-3 flex items-center justify-between text-sm"
            >
              <div
                className="w-0 flex-1 flex items-center"
              >
                <a
                  className="font-medium text-indigo-600"
                  onClick={[Function]}
                >
                  <span
                    className="ml-2"
                  >
                    Mike Erhmantraut
                  </span>
                </a>
              </div>
            </li>
            <li
              className="pl-3 pr-4 py-3 flex items-center justify-between text-sm"
            >
              <div
                className="w-0 flex-1 flex items-center"
              >
                <a
                  className="font-medium text-indigo-600"
                  onClick={[Function]}
                >
                  <span
                    className="ml-2"
                  >
                    Kim Wexler
                  </span>
                </a>
              </div>
            </li>
            <li
              className="pl-3 pr-4 py-3 flex items-center justify-between text-sm"
            >
              <div
                className="w-0 flex-1 flex items-center"
              >
                <a
                  className="font-medium text-indigo-600"
                  onClick={[Function]}
                >
                  <span
                    className="ml-2"
                  >
                    Howard Hamlin
                  </span>
                </a>
              </div>
            </li>
            <li
              className="pl-3 pr-4 py-3 flex items-center justify-between text-sm"
            >
              <div
                className="w-0 flex-1 flex items-center"
              >
                <a
                  className="font-medium text-indigo-600"
                  onClick={[Function]}
                >
                  <span
                    className="ml-2"
                  >
                    Chuck McGill
                  </span>
                </a>
              </div>
            </li>
            <li
              className="pl-3 pr-4 py-3 flex items-center justify-between text-sm"
            >
              <div
                className="w-0 flex-1 flex items-center"
              >
                <a
                  className="font-medium text-indigo-600"
                  onClick={[Function]}
                >
                  <span
                    className="ml-2"
                  >
                    Nacho Varga
                  </span>
                </a>
              </div>
            </li>
          </ul>
        </dd>
      </div>
    </dl>
  </div>
</div>
`;

exports[`EpisodePage must render loading 1`] = `"Loading in progress"`;
